feat: add ability to map model names in the URLS/JSON response

lsmith77 opened the pull request:
implements https://github.com/zenstackhq/zenstack/issues/2178

this PR also adds support for a `prefix` similar to the openapi plugin. note openapi plugin itself still needs to be updated, to also use this model name mapping.

also we did not add any new tests. would appreciate any hints on that.

finally we will of course also be happy to provide documentation for this.

/cc @tobiasbrugger

lsmith77 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -220,29 +237,60 @@ class RequestHandler extends APIHandlerBase {
+    private buildUrlPatternMap(urlSegmentNameCharset: string): Record<UrlPatterns, UrlPattern> {
         const options = { segmentValueCharset: urlSegmentNameCharset };
+
+        const buildPath = (segments: string[]) => {
```
if you prefer we can also drop this change and use `string` rather than `string[]` in the constructor of `UrlPattern`.

ymc9 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -220,29 +236,60 @@ class RequestHandler extends APIHandlerBase {
+
+        const match = pattern.match(path);
+        if (match) {
+            match.type = this.modelNameMapping[match.type] ?? match.type;
```
When a mapping is provided,  say "users" => "user", should access via unmapped path name like "POST /api/user" fail? I tend to think so - configuring a mapping requires you to only use the mapped name.

What do you think? Currently I believe both are valid.

lsmith77 replied:
Indeed. It should fail. We only had the case in mind where only some of the models are mapped but not all.

So should we rather throw an exception if a model is not mapped if there are any mappings?

lsmith77 replied:
I now did some testing. At least with Nest.js it already returns a 404, since there is no `/user` route defined. Now if someone would then also define a `/user` route, indeed without change it would then work but at that point that seems to be intended by the user. So imho there is no need to handle this case inside the `RequestHandler`.

ymc9 replied:
My understanding is the mapping (in the latest code) is inverted.

I tested the latest code with this mapping:

```ts
{
    user: 'myUser',
    post: 'myPost',
}
```

And this test is failing because of returning 200:
```ts
            await expect(
                handler({
                    method: 'post',
                    path: '/user',
                    query: {},
                    requestBody: { data: { type: 'user', attributes: { name: 'User1' } } },
                    prisma,
                })
            ).resolves.toMatchObject({
                status: 400,
            });
```

ymc9 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -220,29 +237,60 @@ class RequestHandler extends APIHandlerBase {
+
+        const match = pattern.match(path);
+        if (match) {
+            match.type = this.modelNameMapping[match.type] ?? match.type;
+            match.relationship = this.modelNameMapping[match.relationship] ?? match.relationship;
+        }
```
@lsmith77 I think @coderabbitai 's comment is valid. The "relationship" part in the URI represents model field names instead of model names, so they shouldn't be mapped.

E.g., if we have a mapped model name "myPost" => "post", we'll still be accessing "/api/user/relationships/posts" without altering the "posts" part, since it's a relation field name.

ymc9 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -531,11 +574,13 @@ class RequestHandler extends APIHandlerBase {
         }
 
         if (entity?.[relationship]) {
+            const mappedType = this.reverseModelNameMap(type);
+            const mappedRelationship = this.reverseModelNameMap(relationship);
```
Same here @lsmith77

ymc9 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -582,11 +627,13 @@ class RequestHandler extends APIHandlerBase {
         }
 
         const entity: any = await prisma[type].findUnique(args);
+        const mappedType = this.reverseModelNameMap(type);
+        const mappedRelationship = this.reverseModelNameMap(relationship);
```
And here

ymc9 commented on packages/server/src/api/rest/index.ts:
```diff
@@ -50,6 +50,9 @@ export type Options = {
     urlSegmentCharset?: string;
+
+    modelNameMapping?: Record<string, string>;
+    prefix?: string;
```
Hi @lsmith77 , I think I'm still not understanding why this is necessary. Isn't configuring a prefix here (e.g., "/v1") equivalent to mounting the API handler directly under the "/v1" route at the framework level?

lsmith77 replied:
In our tests if I just add the `prefix` on the `endpoint`, we would get an error since the regular expressions expect the path to start with model name.

ie. it was trying to match `/api/v1/users` to `/users`

ymc9 replied:
I see. The server adapter that wraps around the api handler is supposed to strip the url prefix (either via configuration or automatically) and pass the clean result to the api handler.

Which server adapter are you using? Is it NestJS? Have you tried the `baseUrl` option? https://zenstack.dev/docs/reference/server-adapters/nestjs#parameter-options-1

lsmith77 replied:
ah indeed, `baseUrl` does the trick. so I will remove the `prefix`. it is a bit unfortunate that in the openapi plugin it is called `prefix` and here it is called `baseUrl`. maybe something to consider for v3

ymc9 reviewed (commented):
Hi @lsmith77, thanks for working on this!

I've put a few review comments there, please take a look when you have time. I've also pushed a sample test case that you can use as a base to expand.

ymc9 commented:
> Hi @lsmith77, thanks for working on this!
> 
> I've put a few review comments there, please take a look when you have time. I've also pushed a sample test case that you can use as a base to expand.

It seems I can't push the test case commit directly. I've put it as a PR here: https://github.com/digital-sustainability/zenstack/pull/1

lsmith77 commented:
> > Hi @lsmith77, thanks for working on this!
> > I've put a few review comments there, please take a look when you have time. I've also pushed a sample test case that you can use as a base to expand.
> 
> It seems I can't push the test case commit directly. I've put it as a PR here: [digital-sustainability#1](https://github.com/digital-sustainability/zenstack/pull/1)

I added one comment. Shouldn't using the unmapped model name cause a 404 and not a 400?

lsmith77 commented on packages/plugins/openapi/src/rest-generator.ts:
```diff
@@ -126,6 +129,10 @@ export class RESTfulOpenAPIGenerator extends OpenAPIGeneratorBase {
     }
 
+    private mapModelName(modelName: string): string {
+        return this.modelNameMapping[modelName] ?? modelName;
```
@ymc9 should we lowercase the `modelName` here?

ymc9 replied:
I think it depends on how we want the user to configure the mapping.

I feel it can be quite confusing what the mapping actually means. Should we make it more explicit and non-ambiguous, and let it represent a mapping from model name (without any casing changes) to resource (url path) name?

```ts
// left side is model name as defined in zmodel
// right side is rest resource name as represented in url and payload
modelResourceNameMap: {
    User: 'users',
    Post: 'posts'
}
```

So that the direction of mapping and the meaning of each side is very obvious and clear?

lsmith77 commented:
@ymc9 is there anything else you need from us before you can merge this?

ymc9 commented:
> @ymc9 is there anything else you need from us before you can merge this?

Hi @lsmith77, it seems the newly added test case is still failing. Also I believe the mapping direction in the test case should be updated based on the latest changes. Could you help fix that, then I think we can merge it. I'll make a follow up PR to add more test cases.

lsmith77 commented:
@ymc9 

> > @ymc9 is there anything else you need from us before you can merge this?
> 
> Hi @lsmith77, it seems the newly added test case is still failing. Also I believe the mapping direction in the test case should be updated based on the latest changes. Could you help fix that, then I think we can merge it. I'll make a follow up PR to add more test cases.

Ah yes, sorry. I didn't take ownership of the test within the PR. I have updated the mapping. I have also now added explicit code to check for the old path (which IMHO should be done by the framework's router and not really here).

But still I am unable to get the tests to pass. Maybe I am running the tests incorrectly.

What I did was:

```
pnpm build
cd packages/server
pnpm test rest.test.ts -- --testNamePattern="REST server tests - model name mapping"
```

ymc9 commented:
> @ymc9
> 
> > > @ymc9 is there anything else you need from us before you can merge this?
> > 
> > 
> > Hi @lsmith77, it seems the newly added test case is still failing. Also I believe the mapping direction in the test case should be updated based on the latest changes. Could you help fix that, then I think we can merge it. I'll make a follow up PR to add more test cases.
> 
> Ah yes, sorry. I didn't take ownership of the test within the PR. I have updated the mapping. I have also now added explicit code to check for the old path (which IMHO should be done by the framework's router and not really here).
> 
> But still I am unable to get the tests to pass. Maybe I am running the tests incorrectly.
> 
> What I did was:
> 
> ```
> pnpm build
> cd packages/server
> pnpm test rest.test.ts -- --testNamePattern="REST server tests - model name mapping"
> ```
> 
> Which results in:
> 
> ```
>    - Expected  - 1
>     + Received  + 1
> 
>       Object {
>     -   "status": 400,
>     +   "status": 201,
>       }
> 
>       3057 |                     prisma,
>       3058 |                 })
>     > 3059 |             ).resolves.toMatchObject({
>            |                        ^
>       3060 |                 status: 400,
>       3061 |             });
>       3062 |
> 
>       at Object.toMatchObject (../../node_modules/expect/build/index.js:174:22)
>       at Object.<anonymous> (tests/api/rest.test.ts:3059:24)
> ```

That's what I meant in a previous comment: even with the mapping "user" -> "myUser", access to route "/user" is still valid given the current implementation, which should actually fail. If a model name is mapped, only the mapped route is accessible. What do you think?

lsmith77 commented:
This aspect I do not understand.

1) There should not be a `/user` route defined within the framework. So the framework should already return a 404. This is at least how it is in Nest.js
2) I have now even added explicit code, which should result in no match being found.

Now for 2), I guess maybe that logic is not sufficient and rather it should trigger an error response directly.

But due to 1), I am not even convinced this case should even be handled explicitly at all.

ymc9 commented:
> This aspect I do not understand.
> 
> 1. There should not be a `/user` route defined within the framework. So the framework should already return a 404. This is at least how it is in Nest.js
> 2. I have now even added explicit code, which should result in no match being found.
> 
> Now for 2), I guess maybe that logic is not sufficient and rather it should trigger an error response directly.
> 
> But due to 1), I am not even convinced this case should even be handled explicitly at all.

I think I understand where the gap is 😄.

The rest api handler is framework agnostic and is supposed to work with different frameworks. For Nest.js, people would explicitly define "users" route, however for other frameworks, e.g., Express.js, one would simply mount the api handler to a parent route and let it handle whatever request sent to it:

```ts
// forward all route under "/api/model" to the underlying request handler
app.use(
    '/api/model',
    ZenStackMiddleware({
        ...
    })
);
```

That's why I think the api handler itself should have a consistent handling of mapped model names, regardless the outer framework rejects an invalid route or not.

ymc9 commented:
> > This aspect I do not understand.
> > 
> > 1. There should not be a `/user` route defined within the framework. So the framework should already return a 404. This is at least how it is in Nest.js
> > 2. I have now even added explicit code, which should result in no match being found.
> > 
> > Now for 2), I guess maybe that logic is not sufficient and rather it should trigger an error response directly.
> > But due to 1), I am not even convinced this case should even be handled explicitly at all.
> 
> I think I understand where the gap is 😄.
> 
> The rest api handler is framework agnostic and is supposed to work with different frameworks. For Nest.js, people would explicitly define "users" route, however for other frameworks, e.g., Express.js, one would simply mount the api handler to a parent route and let it handle whatever request sent to it:
> 
> ```ts
> // forward all route under "/api/model" to the underlying request handler
> app.use(
>     '/api/model',
>     ZenStackMiddleware({
>         ...
>     })
> );
> ```
> 
> That's why I think the api handler itself should have a consistent handling of mapped model names, regardless the outer framework rejects an invalid route or not.

The relationship between framework and api handler is depicted here:

https://zenstack.dev/docs/reference/server-adapters/api-handlers/

lsmith77 commented:
OK, I am now throwing an exception in this case and the tests are now passing.

ymc9 reviewed (approved):
Thanks @lsmith77 ! I'm merging the change and will publish a release soon.
